File: simpat_revision/graphical_abstract_simpat.py
#!/usr/bin/env python3
"""Publication-style graphical abstract for the SIMPAT submission.

The layout is deliberately data-led: no rounded cards, shadows, gradients, or
decorative icons.  PDF and SVG retain vector geometry and editable text.
"""

from pathlib import Path

import matplotlib

matplotlib.use("Agg")
import matplotlib.pyplot as plt
from matplotlib.patches import FancyArrowPatch, Rectangle


HERE = Path(__file__).resolve().parent

# Okabe--Ito-derived, print-safe palette.
BLUE = "#0072B2"
BLUE_LIGHT = "#DCECF4"
ORANGE = "#D55E00"
ORANGE_LIGHT = "#F7E6D8"
GREY_1 = "#F2F2F2"
GREY_2 = "#B8B8B8"
GREY_3 = "#666666"
INK = "#1B1B1B"

plt.rcParams.update(
    {
        "font.family": "serif",
        "font.serif": ["Times New Roman", "DejaVu Serif"],
        "mathtext.fontset": "stix",
        "font.size": 9.0,
        "axes.linewidth": 0.8,
        "pdf.fonttype": 42,
        "ps.fonttype": 42,
        "svg.fonttype": "none",
    }
)


def text(ax, x, y, value, *, size=9.0, weight="normal", color=INK,
         ha="left", va="center", **kwargs):
    """Consistent text helper in normalized figure coordinates."""
    return ax.text(
        x,
        y,
        value,
        fontsize=size,
        fontweight=weight,
        color=color,
        ha=ha,
        va=va,
        **kwargs,
    )


def arrow(ax, start, end, *, color=INK, lw=1.05, scale=10):
    ax.add_patch(
        FancyArrowPatch(
            start,
            end,
            arrowstyle="-|>",
            mutation_scale=scale,
            linewidth=lw,
            color=color,
            shrinkA=0,
            shrinkB=0,
            connectionstyle="arc3,rad=0",
        )
    )


def panel_heading(ax, x, y, letter, heading):
    text(ax, x, y, letter, size=12.5, weight="bold", va="top")
    text(ax, x + 0.025, y - 0.002, heading, size=11.5, weight="bold", va="top")


fig, ax = plt.subplots(figsize=(13.2, 5.25), facecolor="white")
ax.set_xlim(0, 1)
ax.set_ylim(0, 1)
ax.axis("off")

# ---------------------------------------------------------------------------
# Header: modest, left aligned, and separated from the content by one rule.
text(
    ax,
    0.035,
    0.956,
    "Reference-calibrated selective verification of surrogate Jacobians",
    size=16.2,
    weight="bold",
    va="top",
)
text(
    ax,
    0.035,
    0.904,
    "Rank components after training; spend reference simulations only where calibrated evidence is insufficient.",
    size=9.7,
    color=GREY_3,
    va="top",
)
ax.plot([0.035, 0.965], [0.862, 0.862], color=INK, lw=0.85)

# Three restrained columns, separated by hairlines instead of cards.
ax.plot([0.327, 0.327], [0.265, 0.825], color=GREY_2, lw=0.65)
ax.plot([0.698, 0.698], [0.265, 0.825], color=GREY_2, lw=0.65)

# ---------------------------------------------------------------------------
# (a) Cross-retrain evidence: a Jacobian sign matrix and explicit score set.
panel_heading(ax, 0.035, 0.824, "a", "Cross-retrain evidence")
text(ax, 0.060, 0.758, r"Autodiff components from $M$ independently trained models", size=8.8)
text(ax, 0.293, 0.758, r"$k$", size=8.4, color=GREY_3, ha="right")

signs = [
    [1, 1, 1, -1, 1, -1],
    [1, 1, -1, -1, 1, 1],
    [1, 1, 1, -1, -1, -1],
    [1, 1, -1, -1, 1, -1],
    [1, 1, 1, -1, 1, 1],
]
x0, y0 = 0.083, 0.465
cw, ch, gx, gy = 0.028, 0.043, 0.006, 0.009
for row, values in enumerate(signs):
    yy = y0 + (len(signs) - 1 - row) * (ch + gy)
    text(ax, 0.067, yy + ch / 2, rf"{row + 1}", size=7.4, color=GREY_3, ha="right")
    for col, value in enumerate(values):
        xx = x0 + col * (cw + gx)
        face = BLUE_LIGHT if value > 0 else ORANGE_LIGHT
        edge = BLUE if value > 0 else ORANGE
        ax.add_patch(Rectangle((xx, yy), cw, ch, facecolor=face, edgecolor=edge, lw=0.75))
        text(
            ax,
            xx + cw / 2,
            yy + ch / 2,
            "+" if value > 0 else "−",
            size=8.7,
            weight="bold",
            color=edge,
            ha="center",
        )

for col in range(6):
    xx = x0 + col * (cw + gx) + cw / 2
    text(ax, xx, y0 - 0.025, str(col + 1), size=7.4, color=GREY_3, ha="center")
text(ax, 0.047, y0 + 2.5 * (ch + gy) + 0.008, r"$m$", size=8.4, color=GREY_3, ha="right")

ax.plot([0.060, 0.292], [0.420, 0.420], color=GREY_2, lw=0.55)
text(ax, 0.060, 0.386, "Candidate component scores", size=8.6, weight="bold")
text(ax, 0.060, 0.344, r"mean-aligned sign   $a_k$", size=8.7, color=BLUE)
text(ax, 0.060, 0.306, r"signal-to-noise       $u_k=|\overline{J}_k|/\widehat{\sigma}_k$", size=8.7)
text(ax, 0.060, 0.268, r"magnitude               $m_k=|\overline{J}_k|$", size=8.7)

# ---------------------------------------------------------------------------
# (b) Calibration and deployment: threshold axis, branch, and hybrid vector.
panel_heading(ax, 0.350, 0.824, "b", "Calibrate, then deploy")
text(ax, 0.375, 0.758, "Reference-labelled calibration queries", size=9.2, weight="bold")
text(
    ax,
    0.375,
    0.720,
    r"Choose score and $\tau_s$ for the intended loss; abstain if the risk screen fails.",
    size=8.4,
    color=GREY_3,
)

# Score axis for a new design query.
axis_y, axis_x0, axis_x1 = 0.615, 0.385, 0.655
ax.plot([axis_x0, axis_x1], [axis_y, axis_y], color=INK, lw=0.8)
for xx, lab in [(axis_x0, "low"), (axis_x1, "high")]:
    ax.plot([xx, xx], [axis_y - 0.010, axis_y + 0.010], color=INK, lw=0.75)
    text(ax, xx, axis_y - 0.032, lab, size=7.2, color=GREY_3, ha="center", va="top")
tau_x = 0.550
ax.plot([tau_x, tau_x], [axis_y - 0.037, axis_y + 0.047], color=ORANGE, lw=1.15)
text(ax, tau_x, axis_y + 0.060, r"$\tau_s$", size=9.4, color=ORANGE, ha="center")
for xx, marker, face in [
    (0.410, "o", "white"),
    (0.455, "o", ORANGE),
    (0.505, "o", "white"),
    (0.578, "o", BLUE),
    (0.620, "o", BLUE),
    (0.646, "o", "white"),
]:
    ax.scatter([xx], [axis_y], s=30, marker=marker, facecolor=face, edgecolor=INK, linewidth=0.65, zorder=4)
text(ax, 0.500, 0.675, r"deployment scores $s_k$", size=8.2, color=GREY_3, ha="center")

# Straight branch: below threshold is checked; above threshold is retained.
text(ax, 0.424, 0.525, r"$s_k<\tau_s$", size=8.8, weight="bold", color=ORANGE, ha="center")
text(ax, 0.612, 0.525, r"$s_k\geq\tau_s$", size=8.8, weight="bold", color=BLUE, ha="center")
arrow(ax, (0.455, 0.586), (0.430, 0.505), color=ORANGE)
arrow(ax, (0.610, 0.586), (0.610, 0.505), color=BLUE)

ax.add_patch(Rectangle((0.374, 0.405), 0.112, 0.073, facecolor="white", edgecolor=ORANGE, lw=1.0))
text(ax, 0.430, 0.447, "central reference FD", size=9.0, weight="bold", color=ORANGE, ha="center")
text(ax, 0.430, 0.421, "2 simulator evaluations", size=7.4, color=GREY_3, ha="center")
ax.add_patch(Rectangle((0.554, 0.405), 0.112, 0.073, facecolor="white", edgecolor=BLUE, lw=1.0))
text(ax, 0.610, 0.447, r"retain $\overline{J}_k$", size=9.2, weight="bold", color=BLUE, ha="center")
text(ax, 0.610, 0.421, "no simulator call", size=7.4, color=GREY_3, ha="center")

arrow(ax, (0.430, 0.398), (0.486, 0.345), color=ORANGE)
arrow(ax, (0.610, 0.398), (0.553, 0.345), color=BLUE)
text(ax, 0.520, 0.350, "assembled hybrid Jacobian", size=8.7, weight="bold", ha="center", va="bottom")
vx0, vy, vw, vh = 0.437, 0.282, 0.028, 0.047
hybrid = [BLUE, ORANGE, ORANGE, BLUE, BLUE, ORANGE]
for i, col in enumerate(hybrid):
    xx = vx0 + i * (vw + 0.006)
    ax.add_patch(Rectangle((xx, vy), vw, vh, facecolor="white", edgecolor=col, lw=1.2))
    text(ax, xx + vw / 2, vy + vh / 2, r"$\bar J$" if col == BLUE else r"$J^r$",
         size=7.2, color=col, ha="center")

# ---------------------------------------------------------------------------
# (c) Auditable operating point: actual held-out values, not an icon panel.
panel_heading(ax, 0.721, 0.824, "c", r"Measured operating point ($a_k\geq0.9$)")
text(ax, 0.746, 0.752, "benchmark", size=7.7, weight="bold", color=GREY_3)
text(ax, 0.863, 0.752, "coverage", size=7.3, weight="bold", color=GREY_3, ha="center")
text(ax, 0.913, 0.752, "risk", size=7.3, weight="bold", color=GREY_3, ha="center")
text(ax, 0.957, 0.752, "evals/query", size=6.8, weight="bold", color=GREY_3, ha="right")
ax.plot([0.746, 0.958], [0.730, 0.730], color=INK, lw=0.65)

rows = [
    ("TMM near-saturated", "95.9%", "1.5%", "0.90/22"),
    ("TMM stressed", "74.0%", "13.4%", "5.72/22"),
    ("Heat/Poisson stressed", "79.9%", "4.9%", "6.44/32"),
]
for i, (name, cov, risk, cost) in enumerate(rows):
    yy = 0.686 - i * 0.080
    text(ax, 0.746, yy, name, size=8.0)
    text(ax, 0.863, yy, cov, size=8.1, color=BLUE, weight="bold", ha="center")
    text(ax, 0.913, yy, risk, size=8.1, color=ORANGE, weight="bold", ha="center")
    text(ax, 0.957, yy, cost, size=8.1, ha="right")
    ax.plot([0.746, 0.958], [yy - 0.035, yy - 0.035], color=GREY_1, lw=1.0)

text(ax, 0.746, 0.438, "Score choice depends on the loss", size=9.2, weight="bold")
ax.plot([0.746, 0.958], [0.418, 0.418], color=GREY_2, lw=0.55)
text(ax, 0.746, 0.382, "ROC AUC", size=8.0, color=GREY_3)
text(ax, 0.958, 0.382, "SNR: both TMM; magnitude: heat", size=7.8, color=BLUE, ha="right")
text(ax, 0.746, 0.340, "fixed-budget error", size=8.0, color=GREY_3)
text(ax, 0.958, 0.340, "sign: lowest in all three", size=8.0, color=ORANGE, ha="right")
text(ax, 0.746, 0.296, r"$R_{\rm FT}=P(\mathrm{sign\ wrong}\mid\mathrm{accepted})$", size=7.5, color=GREY_3)
text(ax, 0.958, 0.267, "cost column: selective / complete central FD", size=6.8, color=GREY_3, ha="right")

# ---------------------------------------------------------------------------
# One compact scope boundary replaces the former decorative warning banner.
ax.plot([0.035, 0.965], [0.215, 0.215], color=ORANGE, lw=1.15)
text(ax, 0.035, 0.170, "Scope boundary", size=9.2, weight="bold", color=ORANGE)
text(
    ax,
    0.138,
    0.170,
    r"common positive rescaling of all members at fixed $k$ $\Rightarrow$ unchanged sign agreement and SNR; $m_k$ changes",
    size=8.8,
)
text(ax, 0.680, 0.170, r"shared bias $\Rightarrow$ high agreement can still be wrong", size=8.8)
text(
    ax,
    0.035,
    0.105,
    "Evidence screen: 3,580/3,580 external labels passed step/grid checks; unresolved antenna derivatives were excluded.",
    size=8.4,
    color=GREY_3,
)
text(ax, 0.965, 0.105, "calibrated triage  ≠  correctness certificate", size=8.4,
     color=GREY_3, ha="right")

fig.subplots_adjust(left=0, right=1, top=1, bottom=0)
for ext in ("pdf", "svg", "png"):
    fig.savefig(
        HERE / f"graphical_abstract_simpat.{ext}",
        dpi=600 if ext == "png" else None,
        facecolor="white",
        bbox_inches="tight",
        pad_inches=0.04,
    )
print("wrote graphical_abstract_simpat.pdf/svg/png")

File: simpat_revision/graphical_abstract_simpat.R
#!/usr/bin/env Rscript

# Deterministic vector graphical abstract for the SIMPAT submission.
# One R/grid backend produces every preview and export. The design uses only
# square engineering boxes, orthogonal connectors, traceable values, and the
# manuscript-wide blue/grey/warm palette.

suppressPackageStartupMessages({
  library(grid)
  library(svglite)
  library(ragg)
})

args <- commandArgs(trailingOnly = FALSE)
file_arg <- grep("^--file=", args, value = TRUE)
here <- if (length(file_arg)) {
  dirname(normalizePath(sub("^--file=", "", file_arg[1]), winslash = "/"))
} else {
  normalizePath(getwd(), winslash = "/")
}

operating <- read.csv(file.path(here, "graphical_abstract_operating_points.csv"),
                      check.names = FALSE, stringsAsFactors = FALSE)
signs_df <- read.csv(file.path(here, "graphical_abstract_sign_matrix.csv"),
                     check.names = FALSE)
signs <- as.matrix(signs_df[, -1])

W_MM <- 183
H_MM <- 73
W_IN <- W_MM / 25.4
H_IN <- H_MM / 25.4
FONT <- "Times New Roman"

COL <- list(
  blue = "#3775BA",
  blue_mid = "#6B9AC4",
  blue_light = "#DCE8F1",
  orange = "#C76B3C",
  orange_light = "#F5E7DF",
  grey_fill = "#F2F2F2",
  grey_line = "#B8B8B8",
  grey_text = "#666666",
  dark = "#4D4D4D",
  ink = "#1B1B1B",
  white = "#FFFFFF"
)

boxes <- new.env(parent = emptyenv())

txt <- function(x, y, label, size = 6, face = "plain", col = COL$ink,
                just = c("left", "centre"), rot = 0, lineheight = 0.95) {
  grid.text(label, x = unit(x, "npc"), y = unit(y, "npc"),
            just = just, rot = rot,
            gp = gpar(fontfamily = FONT, fontsize = size, fontface = face,
                      col = col, lineheight = lineheight))
}

rule <- function(x0, y0, x1, y1, col = COL$grey_line, lwd = 0.7) {
  grid.lines(unit(c(x0, x1), "npc"), unit(c(y0, y1), "npc"),
             gp = gpar(col = col, lwd = lwd, lineend = "butt"))
}

rect_node <- function(id, x0, y0, x1, y1, fill = COL$white,
                      stroke = COL$dark, lwd = 0.8) {
  grid.rect(x = unit((x0 + x1) / 2, "npc"), y = unit((y0 + y1) / 2, "npc"),
            width = unit(x1 - x0, "npc"), height = unit(y1 - y0, "npc"),
            gp = gpar(fill = fill, col = stroke, lwd = lwd, linejoin = "mitre"))
  assign(id, c(x0 = x0, y0 = y0, x1 = x1, y1 = y1), envir = boxes)
}

on_boundary <- function(p, b, tol = 1e-8) {
  on_x <- (abs(p[1] - b["x0"]) < tol || abs(p[1] - b["x1"]) < tol) &&
    p[2] >= b["y0"] - tol && p[2] <= b["y1"] + tol
  on_y <- (abs(p[2] - b["y0"]) < tol || abs(p[2] - b["y1"]) < tol) &&
    p[1] >= b["x0"] - tol && p[1] <= b["x1"] + tol
  on_x || on_y
}

inside_strict <- function(x, y, b, tol = 1e-6) {
  x > b["x0"] + tol && x < b["x1"] - tol &&
    y > b["y0"] + tol && y < b["y1"] - tol
}

validate_route <- function(points, start_id = NULL, end_id = NULL) {
  stopifnot(ncol(points) == 2, nrow(points) >= 2)
  d <- diff(points)
  if (any(abs(d[, 1]) > 1e-10 & abs(d[, 2]) > 1e-10))
    stop("Connector contains a non-orthogonal segment")
  if (!is.null(start_id) && !on_boundary(points[1, ], get(start_id, boxes)))
    stop("Connector does not start on the source-node boundary: ", start_id)
  if (!is.null(end_id) && !on_boundary(points[nrow(points), ], get(end_id, boxes)))
    stop("Connector does not end on the target-node boundary: ", end_id)

  ids <- ls(boxes)
  for (i in seq_len(nrow(points) - 1)) {
    t <- seq(0.02, 0.98, length.out = 60)
    xs <- points[i, 1] + t * (points[i + 1, 1] - points[i, 1])
    ys <- points[i, 2] + t * (points[i + 1, 2] - points[i, 2])
    for (id in ids) {
      if (id %in% c(start_id, end_id)) next
      b <- get(id, boxes)
      if (any(mapply(inside_strict, xs, ys, MoreArgs = list(b = b))))
        stop("Connector intersects node interior: ", id)
    }
  }
  invisible(TRUE)
}

ortho_arrow <- function(points, start_id = NULL, end_id = NULL,
                        col = COL$dark, lwd = 0.9, head_mm = 1.8) {
  points <- as.matrix(points)
  validate_route(points, start_id, end_id)
  grid.lines(unit(points[, 1], "npc"), unit(points[, 2], "npc"),
             arrow = arrow(type = "closed", ends = "last",
                           length = unit(head_mm, "mm")),
             gp = gpar(col = col, fill = col, lwd = lwd,
                       lineend = "butt", linejoin = "mitre"))
}

stage_heading <- function(x, y, number, label) {
  grid.rect(unit(x, "npc"), unit(y, "npc"),
            width = unit(0.022, "npc"), height = unit(0.030, "npc"),
            gp = gpar(fill = COL$blue, col = COL$blue, lwd = 0.6))
  txt(x, y, number, size = 6.2, face = "bold", col = COL$white,
      just = c("centre", "centre"))
  txt(x + 0.018, y, label, size = 7.2, face = "bold")
}

draw_figure <- function() {
  grid.newpage()
  pushViewport(viewport(xscale = c(0, 1), yscale = c(0, 1),
                        gp = gpar(fontfamily = FONT)))

  # Header
  txt(0.026, 0.950, "Reference-calibrated verification of metamodel design Jacobians",
      size = 12.2, face = "bold", just = c("left", "top"))
  txt(0.026, 0.898,
      "Rank components, verify only where calibrated evidence is insufficient, and report empirical risk at the actual simulation cost.",
      size = 6.6, col = COL$grey_text, just = c("left", "top"))
  rule(0.026, 0.855, 0.974, 0.855, COL$ink, 0.8)

  # Stage separators establish hierarchy without rounded card layouts.
  rule(0.708, 0.160, 0.708, 0.820, COL$grey_line, 0.65)
  stage_heading(0.040, 0.812, "1", "Cross-retrain evidence")
  stage_heading(0.335, 0.812, "2", "Calibrate and allocate")
  stage_heading(0.733, 0.812, "3", "Report the operating point")

  # Stage 1: deterministic sign matrix from traceable CSV input.
  txt(0.050, 0.755, "Autodiff signs from M independent retrains", size = 5.8)
  x0 <- 0.075; y_top <- 0.700; cw <- 0.029; ch <- 0.040
  gx <- 0.006; gy <- 0.010
  for (r in seq_len(nrow(signs))) {
    yy <- y_top - (r - 1) * (ch + gy)
    txt(0.061, yy, as.character(signs_df$retrain[r]), size = 4.7,
        col = COL$grey_text, just = c("right", "centre"))
    for (k in seq_len(ncol(signs))) {
      xx <- x0 + (k - 1) * (cw + gx)
      positive <- signs[r, k] > 0
      grid.rect(unit(xx, "npc"), unit(yy, "npc"),
                width = unit(cw, "npc"), height = unit(ch, "npc"),
                gp = gpar(fill = if (positive) COL$blue_light else COL$orange_light,
                          col = if (positive) COL$blue else COL$orange,
                          lwd = 0.65))
      txt(xx, yy, if (positive) "+" else "−", size = 6.0, face = "bold",
          col = if (positive) COL$blue else COL$orange,
          just = c("centre", "centre"))
    }
  }
  for (k in seq_len(ncol(signs))) {
    xx <- x0 + (k - 1) * (cw + gx)
    txt(xx, 0.438, paste0("g", k), size = 4.6, col = COL$grey_text,
        just = c("centre", "centre"))
  }
  txt(0.052, 0.570, "retrain", size = 4.7, col = COL$grey_text,
      rot = 90, just = c("centre", "centre"))

  rect_node("scores", 0.050, 0.235, 0.285, 0.380,
            fill = COL$white, stroke = COL$blue, lwd = 0.9)
  txt(0.062, 0.352, "Candidate component scores", size = 5.9, face = "bold")
  txt(0.062, 0.316, "sign agreement  aₖ", size = 5.4, col = COL$blue)
  txt(0.062, 0.282, "signal-to-noise  uₖ", size = 5.4)
  txt(0.205, 0.282, "magnitude  mₖ", size = 5.4)

  # Stage 2: two calibration inputs, one prespecified operating rule.
  rect_node("reference", 0.355, 0.685, 0.662, 0.758,
            fill = COL$grey_fill, stroke = COL$dark, lwd = 0.8)
  txt(0.508, 0.729, "Independent reference-labelled calibration queries",
      size = 5.8, face = "bold", just = c("centre", "centre"))
  txt(0.508, 0.701, "numerically screened central differences",
      size = 4.9, col = COL$grey_text, just = c("centre", "centre"))

  rect_node("loss", 0.388, 0.565, 0.630, 0.637,
            fill = COL$white, stroke = COL$dark, lwd = 0.8)
  txt(0.509, 0.608, "Select the score for the intended loss",
      size = 5.7, face = "bold", just = c("centre", "centre"))
  txt(0.509, 0.581, "ranking AUC  |  risk–coverage  |  budgeted gradient error",
      size = 4.7, col = COL$grey_text, just = c("centre", "centre"))

  rect_node("rule", 0.375, 0.440, 0.643, 0.515,
            fill = COL$blue_light, stroke = COL$blue, lwd = 0.9)
  txt(0.509, 0.484, "Calibrated threshold τₛ or fixed budget B",
      size = 5.8, face = "bold", col = COL$blue,
      just = c("centre", "centre"))
  txt(0.509, 0.456, "verify the B lowest-ranked components",
      size = 4.9, col = COL$dark, just = c("centre", "centre"))

  rect_node("accept", 0.348, 0.280, 0.465, 0.355,
            fill = COL$white, stroke = COL$blue, lwd = 1.0)
  txt(0.4065, 0.327, "ACCEPT", size = 6.0, face = "bold", col = COL$blue,
      just = c("centre", "centre"))
  txt(0.4065, 0.299, "retain mean Jₖ", size = 4.8,
      just = c("centre", "centre"))

  rect_node("verify", 0.548, 0.280, 0.665, 0.355,
            fill = COL$white, stroke = COL$orange, lwd = 1.0)
  txt(0.6065, 0.327, "VERIFY", size = 6.0, face = "bold", col = COL$orange,
      just = c("centre", "centre"))
  txt(0.6065, 0.299, "central reference FD", size = 4.8,
      just = c("centre", "centre"))

  rect_node("hybrid", 0.360, 0.170, 0.540, 0.235,
            fill = COL$white, stroke = COL$dark, lwd = 0.8)
  txt(0.450, 0.221, "hybrid design Jacobian", size = 5.2, face = "bold",
      just = c("centre", "centre"))
  vec_cols <- c(COL$blue, COL$orange, COL$orange, COL$blue, COL$blue, COL$orange)
  for (i in seq_along(vec_cols)) {
    xx <- 0.382 + (i - 1) * 0.027
    grid.rect(unit(xx, "npc"), unit(0.190, "npc"),
              width = unit(0.020, "npc"), height = unit(0.023, "npc"),
              gp = gpar(fill = COL$white, col = vec_cols[i], lwd = 0.85))
  }

  rect_node("abstain", 0.563, 0.170, 0.675, 0.235,
            fill = COL$grey_fill, stroke = COL$grey_text, lwd = 0.8)
  txt(0.619, 0.211, "ABSTAIN / REFINE", size = 5.0, face = "bold",
      col = COL$grey_text, just = c("centre", "centre"))
  txt(0.619, 0.187, "if the screen fails", size = 4.5,
      col = COL$grey_text, just = c("centre", "centre"))

  # Orthogonal routes meet node boundaries and remain clear of labels.
  ortho_arrow(rbind(c(0.285, 0.308), c(0.335, 0.308), c(0.335, 0.601), c(0.388, 0.601)),
              "scores", "loss", COL$blue, 0.85)
  ortho_arrow(rbind(c(0.508, 0.685), c(0.508, 0.637)),
              "reference", "loss", COL$dark, 0.8)
  ortho_arrow(rbind(c(0.509, 0.565), c(0.509, 0.515)),
              "loss", "rule", COL$dark, 0.8)
  ortho_arrow(rbind(c(0.445, 0.440), c(0.445, 0.397), c(0.4065, 0.397), c(0.4065, 0.355)),
              "rule", "accept", COL$blue, 0.9)
  ortho_arrow(rbind(c(0.575, 0.440), c(0.575, 0.397), c(0.6065, 0.397), c(0.6065, 0.355)),
              "rule", "verify", COL$orange, 0.9)
  txt(0.4065, 0.408, "reliable", size = 4.6, col = COL$blue,
      just = c("centre", "centre"))
  txt(0.6065, 0.408, "least reliable", size = 4.6, col = COL$orange,
      just = c("centre", "centre"))
  ortho_arrow(rbind(c(0.4065, 0.280), c(0.4065, 0.235)),
              "accept", "hybrid", COL$blue, 0.9)
  ortho_arrow(rbind(c(0.6065, 0.280), c(0.6065, 0.255), c(0.500, 0.255), c(0.500, 0.235)),
              "verify", "hybrid", COL$orange, 0.9)
  ortho_arrow(rbind(c(0.665, 0.318), c(0.689, 0.318), c(0.689, 0.2025), c(0.675, 0.2025)),
              "verify", "abstain", COL$grey_text, 0.75)

  # Stage 3: measured external operating points, loaded from CSV.
  txt(0.744, 0.757, "External regimes at τ = 0.9",
      size = 5.8, face = "bold")
  txt(0.744, 0.710, "benchmark", size = 4.7, face = "bold", col = COL$grey_text)
  txt(0.873, 0.710, "coverage", size = 4.6, face = "bold", col = COL$grey_text,
      just = c("centre", "centre"))
  txt(0.917, 0.710, "risk", size = 4.6, face = "bold", col = COL$grey_text,
      just = c("centre", "centre"))
  txt(0.970, 0.710, "evals", size = 4.5, face = "bold", col = COL$grey_text,
      just = c("right", "centre"))
  rule(0.744, 0.687, 0.970, 0.687, COL$ink, 0.6)

  row_y <- c(0.641, 0.580, 0.519)
  for (i in seq_len(nrow(operating))) {
    txt(0.744, row_y[i], operating$benchmark[i], size = 4.9)
    txt(0.873, row_y[i], sprintf("%.1f%%", operating$coverage_pct[i]),
        size = 5.0, face = "bold", col = COL$blue,
        just = c("centre", "centre"))
    txt(0.917, row_y[i], sprintf("%.1f%%", operating$false_trust_risk_pct[i]),
        size = 5.0, face = "bold", col = COL$orange,
        just = c("centre", "centre"))
    txt(0.970, row_y[i], sprintf("%.2f/%d", operating$selective_evals[i],
                                 operating$complete_fd_evals[i]),
        size = 4.8, just = c("right", "centre"))
    rule(0.744, row_y[i] - 0.031, 0.970, row_y[i] - 0.031,
         COL$grey_fill, 1.0)
  }
  txt(0.970, 0.470, "evals: selective / complete central FD", size = 4.2,
      col = COL$grey_text, just = c("right", "centre"))

  rect_node("loss_report", 0.744, 0.294, 0.970, 0.435,
            fill = COL$white, stroke = COL$grey_line, lwd = 0.7)
  txt(0.756, 0.408, "Score choice follows the deployment loss", size = 5.5, face = "bold")
  txt(0.756, 0.368, "sign-correctness AUC", size = 4.8, col = COL$grey_text)
  txt(0.960, 0.368, "SNR / magnitude", size = 4.9, col = COL$blue,
      face = "bold", just = c("right", "centre"))
  txt(0.756, 0.329, "fixed-budget gradient error", size = 4.8, col = COL$grey_text)
  txt(0.960, 0.329, "sign ordering", size = 4.9, col = COL$orange,
      face = "bold", just = c("right", "centre"))

  rect_node("evidence", 0.744, 0.170, 0.970, 0.252,
            fill = COL$grey_fill, stroke = COL$grey_text, lwd = 0.7)
  txt(0.857, 0.222, "3,580 / 3,580 external labels passed numerical screens",
      size = 4.9, face = "bold", just = c("centre", "centre"))
  txt(0.857, 0.190, "unresolved antenna derivatives were excluded",
      size = 4.6, col = COL$grey_text, just = c("centre", "centre"))

  # Bottom structural boundary: evidence limitation, not decoration.
  rule(0.026, 0.125, 0.974, 0.125, COL$orange, 1.0)
  txt(0.026, 0.082, "Structural boundary", size = 5.7, face = "bold", col = COL$orange)
  txt(0.140, 0.082,
      "positive rescaling leaves sign agreement and SNR unchanged; shared bias can make every retrain agree and still be wrong",
      size = 5.3)
  txt(0.974, 0.042, "calibrated triage  ≠  correctness certificate",
      size = 5.2, face = "bold", col = COL$grey_text,
      just = c("right", "centre"))

  popViewport()
}

base <- file.path(here, "graphical_abstract_simpat")

svglite::svglite(paste0(base, ".svg"), width = W_IN, height = H_IN,
                 bg = "white", system_fonts = list(sans = FONT, serif = FONT))
draw_figure()
dev.off()

grDevices::cairo_pdf(paste0(base, ".pdf"), width = W_IN, height = H_IN,
                     family = FONT, bg = "white")
draw_figure()
dev.off()

ragg::agg_png(paste0(base, ".png"), width = W_IN, height = H_IN,
              units = "in", res = 300, background = "white", scaling = 1)
draw_figure()
dev.off()

ragg::agg_tiff(paste0(base, ".tiff"), width = W_IN, height = H_IN,
               units = "in", res = 600, background = "white",
               compression = "lzw", scaling = 1)
draw_figure()
dev.off()

cat(sprintf("wrote %s.{svg,pdf,png,tiff}\n", base))
